test/unit/container_logs_test.rb

```diff
@@ -0,0 +1,127 @@
+# frozen_string_literal: true
+require 'test_helper'
+
+class ContainerLogsTest < KubernetesDeploy::TestCase
+  def setup
+    super
+    @logs = KubernetesDeploy::ContainerLogs.new(parent_id: 'pod/pod-123-456', container_name: 'A', logger: logger)
+  end
+
+  def test_sync_deduplicates_logs_emitted_fractional_seconds_apart
+    kubectl = mock
+    kubectl.stubs(:run)
+      .returns([logs_part_1, "", ""])
+      .then.returns([logs_part_2, "", ""])
+      .then.returns([logs_part_3, "", ""])
+    @logs.sync(kubectl)
+    @logs.sync(kubectl)
+    @logs.sync(kubectl)
+
+    assert_equal expected_log_lines(1..15), @logs.lines
+  end
+
+  def test_sync_handles_cycles_where_no_new_logs_available
+    kubectl = mock
+    kubectl.stubs(:run)
+      .returns([logs_part_1, "", ""])
+      .then.returns(["", "", ""])
+      .then.returns([logs_part_2, "", ""])
+    @logs.sync(kubectl)
+    @logs.sync(kubectl)
+    @logs.sync(kubectl)
+
+    assert_equal expected_log_lines(1..10), @logs.lines
+  end
+
+  def test_empty_delegated_to_lines
+    kubectl = mock
+    kubectl.stubs(:run).returns([logs_part_1, "", ""])
+    assert_predicate @logs, :empty?
+    @logs.sync(kubectl)
+    refute_predicate @logs, :empty?
+  end
+
+  def test_print_latest_and_print_all_output_the_correct_chunks
+    kubectl = mock
+    kubectl.stubs(:run)
+      .returns([logs_part_1, "", ""])
+      .then.returns([logs_part_2, "", ""])
+
+    @logs.sync(kubectl)
+    @logs.print_latest
+    assert_logs_match_all(expected_log_lines(1..3), in_order: true)
+
+    reset_logger
+    @logs.print_all
+    assert_logs_match_all(expected_log_lines(1..3), in_order: true)
+
+    reset_logger
+    @logs.sync(kubectl)
+    @logs.print_latest
+    assert_logs_match_all(expected_log_lines(4..10), in_order: true)
+    refute_logs_match("Line 3")
+
+    reset_logger
+    @logs.print_all
+    assert_logs_match_all(expected_log_lines(1..10), in_order: true)
+  end
+
+  def test_print_latest_supports_prefixing
+    kubectl = mock
+    kubectl.stubs(:run).returns([logs_part_1, "", ""])
+    @logs.sync(kubectl)
+    expected = [
+      "[A]  Line 1",
+      "[A]  Line 2",
+      "[A]  Line 3"
+    ]
+    @logs.print_latest(prefix: true)
+    assert_logs_match_all(expected, in_order: true)
+  end
+
+  private
+
+  def expected_log_lines(range)
+    range.map { |i| "Line #{i}" }
+  end
+
+  def logs_part_1
```

Comment:
`logs_1_through_3` might be more clear than part_1

Reply:
The chunking isn't based on number of lines though; these three methods are simulating three requests to the API server with sub-second log entries:
- Part 1: requests all logs so far, during second 1. Gets the first few logs from second 1.
- Part 2: requests all logs since second 1, during second 2. Gets all logs from second 1 and the first few from second 2.
- Part 3: requests all logs since second 2, during second 3. Gets all logs from second 2 and the first few from second 3.

Not sure how to find names that explain that succinctly. I could make a fancier helper like `dummy_log_response(start_at_second:, current_second:)` but not sure it's worth the effort.

Reply:
My concern was about the disconnect between `kubectl.stubs(:run).returns([logs_part_1, "", ""])`  and `expected_log_lines(1..3)`